File: design-add-and-search-words-data-structure.cpp
using namespace std;

#define Index(ch) (ch - 'a')

struct TrieNode {
	TrieNode *children[26];
	bool isEnd;

	TrieNode() {
		isEnd = false;
		for (auto &c: children) c = nullptr;
	}

	TrieNode *get_or_create(char ch) {
		int i = Index(ch);
		if (children[i] != nullptr) {
			return children[i];
		}
		TrieNode *node = new TrieNode();
		children[i] = node;
		return node;
	}

	TrieNode *get(char ch) {
		return children[Index(ch)];
	}
};

class WordDictionary {
	TrieNode *root;
public:
    WordDictionary() {
        root = new TrieNode();
    }
    
    void addWord(string word) {
        TrieNode *node = root;
		for (char ch: word) {
			node = node->get_or_create(ch);
		}
		node->isEnd = true;
    }
    
    bool search(string word) {
		return _search(root, word, 0);        
    }

private:
	bool _search(TrieNode *node, string word, int i) {
		if (i >= word.size()) return (node == nullptr || node->isEnd);

		if (word[i] == '.') {
			for (int j=0; j<26; ++j) {
				TrieNode *child = node->children[j];
				if (child == nullptr) continue;
				if (_search(child, word, i+1)) return true;
			}
			return false;
		}

		node = node->get(word[i]);
		return (node == nullptr)? false: _search(node, word, i+1);
	}
};

File: design-add-and-search-words-data-structure.py
from typing import Optional, Self

class TrieNode:
    def __init__(self):
        self.children = {}
        self.end = False

    def get_or_create(self, char: str) -> Self:
        if char in self.children:
            return self.children[char]
        node = TrieNode()
        self.children[char] = node
        return node

    def get(self, char: str) -> Optional[Self]:
        return self.children.get(char, None)

class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root
        for ch in word:
            node = node.get_or_create(ch)
        node.end = True

    def search(self, word: str) -> bool:
        return self._search(self.root, word, 0)

    def _search(self, node: TrieNode, word: str, i: int) -> bool:
        if i >= len(word):
            return node is None or node.end

        ch = word[i]
        if ch == ".":
            for child_node in node.children.values():
                if self._search(child_node, word, i+1):
                    return True
            return False

        node = node.get(ch)
        return False if node is None else self._search(node, word, i+1)
